README
This package has the Vyatta configuration templates and scripts to configure
the IPv6 router advertisement daemon (radvd).


debian/changelog
vyatta-ipv6-rtradv (1.2.6) unstable; urgency=low

  * New release.

 -- VyOS maintainers <maintainers@vyos.io>  Fri, 11 Sep 2020 04:33:35 +0300

vyatta-ipv6-rtradv (1.2.5) unstable; urgency=medium

  * New dnssl option.
  * Fix RA broken on boot.

 -- Daniil Baturin <daniil@vyos.io>  Mon, 09 Mar 2020 13:34:00 +0100

vyatta-ipv6-rtradv (1.2.4) unstable; urgency=low

  * New release.

 -- VyOS maintainers <maintainers@vyos.io>  Tue, 10 Dec 2019 23:40:06 +0200

vyatta-ipv6-rtradv (1.2.0-1) stable; urgency=medium

  * New branch.

 -- Daniil Baturin <daniil@baturin.org>  Fri, 28 Sep 2018 21:58:24 +0200

vyatta-ipv6-rtradv (0.38+vyos2+current1) unstable; urgency=medium

  [ UnicronNL ]
  * Make it compile in debian jessie

  [ Thomas Jepp ]
  * Fix build depends and remove the opt/vyatta/etc directory from debian/vyatta-ipv6-rtradv.install (was causing a build failure).

  [ Kim Hagen ]

 -- Kim Hagen <kim.sidney@gmail.com>  Sun, 24 Jan 2016 15:15:57 -0500

vyatta-ipv6-rtradv (0.38+vyos2+lithium7) unstable; urgency=low

  [ Alex Harpin ]
  * vyatta-ipv6-rtradv: formatting changes for style consistency
  * vyatta-ipv6-rtradv: drop ipv6 templates for unsupported interface types
  * vyatta-ipv6-rtradv: drop the vyatta-ipv6-rtradv-serial package
  * vyatta-ipv6-rtradv: add node priority to template generated interfaces
  
 -- Alex Harpin <development@landsofshadow.co.uk>  Sat, 14 Nov 2015 15:17:24 +0000
 
vyatta-ipv6-rtradv (0.38+vyos2+lithium6) unstable; urgency=low

  [ Patrick van Staveren ]
  * Bug 600: Fix ipv6 autoconf for PPPoE

  [ Daniil Baturin ]

 -- Daniil Baturin <daniil@baturin.org>  Tue, 10 Nov 2015 08:47:56 +0100

vyatta-ipv6-rtradv (0.38+vyos2+lithium5) unstable; urgency=low

  * vyatta-ipv6-rtradv: update dh_gencontrol with new development flag

 -- Alex Harpin <development@landsofshadow.co.uk>  Fri, 19 Jun 2015 07:25:06 +0100
 
vyatta-ipv6-rtradv (0.38+vyos2+lithium4) unstable; urgency=low

  * Update maintainer address

 -- Alex Harpin <development@landsofshadow.co.uk>  Fri, 26 Dec 2014 11:15:01 +0000
 
vyatta-ipv6-rtradv (0.38+vyos2+lithium3) unstable; urgency=low

  * Force release

 -- Alex Harpin <development@landsofshadow.co.uk>  Sun, 21 Dec 2014 22:20:36 +0000
 
vyatta-ipv6-rtradv (0.38+vyos2+lithium2) unstable; urgency=low

  * Force release

 -- Alex Harpin <development@landsofshadow.co.uk>  Tue, 16 Dec 2014 07:48:14 +0000
 
vyatta-ipv6-rtradv (0.38+vyos2+lithium1) unstable; urgency=low

  * New branch

 -- Alex Harpin <development@landsofshadow.co.uk>  Mon, 08 Dec 2014 14:24:48 +0000
 
vyatta-ipv6-rtradv (0.38+vyos1+helium5) unstable; urgency=low

  * Add L2TPv3 and VXLAN to generated templates.

 -- Daniil Baturin <daniil@baturin.org>  Sat, 20 Sep 2014 10:15:14 +0200

vyatta-ipv6-rtradv (0.38+vyos1+helium4) unstable; urgency=low

  * Bug #274: add value completion for IPv6 RA boolean options.

 -- Daniil Baturin <daniil@baturin.org>  Fri, 01 Aug 2014 18:53:06 +0200

vyatta-ipv6-rtradv (0.38+vyos1+helium3) unstable; urgency=low

  * Remove stray dot after brace.

 -- Daniil Baturin <daniil@baturin.org>  Fri, 01 Aug 2014 01:17:06 +0200

vyatta-ipv6-rtradv (0.38+vyos1+helium2) unstable; urgency=low

  [ Ivan Malyarchuk ]
  * add configuration parameters for rfc6106 Recursive DNS Servers

  [ Daniil Baturin ]

 -- Daniil Baturin <daniil@baturin.org>  Wed, 30 Jul 2014 00:29:21 +0200

vyatta-ipv6-rtradv (0.38+vyos1+helium1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sat, 15 Feb 2014 19:16:27 +0100

vyatta-ipv6-rtradv (0.38+hydrogen1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sun, 17 Nov 2013 00:27:21 +0100

vyatta-ipv6-rtradv (0.38+daisy1) unstable; urgency=low

  * create daisy branch

 -- John Southworth <john.southworth@vyatta.com>  Sat, 13 Oct 2012 13:31:03 -0700

vyatta-ipv6-rtradv (0.38) unstable; urgency=low

  * new branch

 -- John Southworth <john.southworth@vyatta.com>  Fri, 12 Oct 2012 19:47:27 -0700

vyatta-ipv6-rtradv (0.37) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Sat, 03 Mar 2012 02:26:00 -0800

vyatta-ipv6-rtradv (0.36) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Thu, 07 Jul 2011 20:57:09 -0700

vyatta-ipv6-rtradv (0.35) unstable; urgency=low

  * Bugfix 6917: Add support for AdvDefaultPreference parameter.

 -- Bob Gilligan <gilligan@vyatta.com>  Sun, 03 Jul 2011 11:41:23 -0700

vyatta-ipv6-rtradv (0.34) unstable; urgency=low

  * Parital fix for bug 6759 serial packages are incorrectly included in
    virt ISO

 -- Mohit Mehta <mohit@vyatta.com>  Wed, 02 Feb 2011 12:24:41 -0800

vyatta-ipv6-rtradv (0.33) unstable; urgency=low

  * Remove unnecessary FileHandle
  * Fix ipv6 parameter path
  * Use perl foreach loop
  * Generate configuration to buffer not temporary file
  * Remove unnecessary dependency on FileHandle

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Wed, 05 Jan 2011 11:52:23 -0800

vyatta-ipv6-rtradv (0.32) unstable; urgency=low

  * new branch

 -- An-Cheng Huang <ancheng@vyatta.com>  Tue, 28 Dec 2010 13:49:29 -0800

vyatta-ipv6-rtradv (0.31) unstable; urgency=low

  * Fix help text

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Mon, 06 Dec 2010 17:08:32 -0800

vyatta-ipv6-rtradv (0.30) unstable; urgency=low

  * Fix missing semicolon error

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Tue, 23 Nov 2010 19:00:48 -0800

vyatta-ipv6-rtradv (0.29) unstable; urgency=low

  * remove deprecated linda override

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Fri, 12 Nov 2010 14:11:24 -0800

vyatta-ipv6-rtradv (0.28) unstable; urgency=low

  * Fixup router advertisement for Mendocino

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Tue, 19 Oct 2010 14:58:53 -0700

vyatta-ipv6-rtradv (0.27) unstable; urgency=low

  * remove old modprobe file during upgrade to mendocino

 -- Mohit Mehta <mohit@vyatta.com>  Mon, 13 Sep 2010 17:11:55 -0700


gen-interface-templates.pl
#!/usr/bin/perl
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2009 Vyatta, Inc.
# All Rights Reserved.
#
# Author: Bob Gilligan
# Date: August 2009
# Description: Script to automatically generate per-interface templates
#              for IPv6 Router Advertisements
#
# **** End License ****

use strict;
use warnings;

# set DEBUG in environment to test script
my $debug = $ENV{'DEBUG'};

# Mapping from configuration level to ifname used AT THAT LEVEL
# Priority values are taken from the relevant node, any changes to those need to be reflected here.
my %interface_hash = (
    'wirelessmodem/node.tag'                        => {'if_name' => '$VAR(@)', 'priority' => '350'},
);

sub gen_template {
    my ($inpath, $outpath, $ifname, $priority) = @_;

    print $outpath, "\n" if ($debug);
    opendir my $d, $inpath
        or die "Can't open: $inpath:$!";

    # walk through sample templates
    foreach my $name (grep {!/^\./} readdir $d) {

        my $in  = "$inpath/$name";
        my $out = "$outpath/$name";

        # recurse into subdirectory
        if (-d $in) {
            my $subif = $ifname;
            $subif =~ s#@\)#../@)#g if ($name ne 'node.tag');

            (-d $out)
                or mkdir($out)
                or die "Can't create $out: $!";

            gen_template($in, $out, $subif, $priority);
            next;
        }

        print "in: $in out: $out\n" if ($debug);
        open my $inf,  '<', $in  or die "Can't open $in: $!";
        open my $outf, '>', $out or die "Can't open $out: $!";

        if ($out =~ /autoconf/) {
            $priority += 2;
        } elsif ( $out =~ /disable-forwarding/) {
            $priority += 1;
        }

        while (my $line = <$inf>) {
            $line =~ s#\$PRIORITY#$priority#;
            $line =~ s#\$IFNAME#$ifname#;
            print $outf $line;
        }
        close $inf;
        close $outf or die "Close error $out:$!";
    }
    closedir $d;
}

sub mkdir_p {
    my $path = shift;

    return 1 if (mkdir($path));

    my $pos = rindex($path, "/");
    return unless $pos != -1;
    return unless mkdir_p(substr($path, 0, $pos));
    return mkdir($path);
}

die "Usage: $0 output_directory\n" if ($#ARGV < 0);

my $outdir = $ARGV[0];

foreach my $if_tree (keys %interface_hash) {
    my $inpath  = "interface-templates";
    my $outpath = "$outdir/interfaces/$if_tree";
    (-d $outpath)
        or mkdir_p($outpath)
        or die "Can't create $outpath:$!";

    gen_template($inpath, $outpath, $interface_hash{$if_tree}{'if_name'}, $interface_hash{$if_tree}{'priority'});
}

# Local Variables:
# mode: perl
# indent-tabs-mode: nil
# perl-indent-level: 4
# End:


scripts/vyatta_gen_radvd.pl
#!/usr/bin/perl
#
# Module: vyatta_gen_radvd.pl
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2005-2009 Vyatta, Inc.
# All Rights Reserved.
#
# Author: Bob Gilligan <gilligan@vyatta.com>
# Date: 2009
# Description:  Setup section of the radvd.conf file for an interface.
#
# **** End License ****
#
#

#
# Syntax:
#    vyatta_gen_radvd.pl --generate <ifname>
#    vyatta_gen_radvd.pl --delete <ifname>
#
# In the first instance, <ifname> is the name of the interface for
# which the configuration file is to be generated (e.g. eth0, eth7.3,
# pppoe7, wan0.1, ml0, etc.) and <config-param-root> is the root of
# the tree of configuration parameters
# (e.g. "/interfaces/ethernet/eth0/ipv6/router-advert").
#
# This script generates a partial radvd.conf file holding the
# parameters for one interface, and writes it to a temp file.  It then
# strips any previous configuration for this interface from the
# system-wide config file /etc/radvd.conf.  Then it cat's the contents
# of the temp file into the system-wide config file /etc/radvd.conf.
# After that it re-starts the radvd daemon.  Since the daemon is stateless,
# re-starting it is a lightweight operation.
#
# In the second instance, <ifname> is the name of the interface
# whose configuration is to be removed from the /etc/radvd.conf file

use strict;
use lib "/opt/vyatta/share/perl5/";

use Vyatta::Config;
use Getopt::Long;
use Vyatta::Interface;

my $conf_file="/etc/radvd.conf";

# Set to 1 to enable debug output
my $debug_flag=0;

sub log_msg {
    my $message = shift;

    if ($debug_flag > 0) {
        print "DEBUG: $message"
    }
}

my $config;
my $FD_WR;      # Temporary config file pointer
my $generate;
my $delete;

GetOptions("generate=s"       => \$generate,
           "delete=s"         => \$delete
);

if ($generate) {
    generate_conf($generate);
    exit 0;
}

if ($delete) {
    delete_conf($delete);
    exit 0;
}

printf("Error: invalid args: $ARGV\n");
exit 1;

# Parse and write config file for one prefix under an interface
sub do_prefix {
    my ($param_root, $prefix) = @_;

    # RFC-4861 parameters for each element of AdvPrefixList.
    # Parameters with fixed default values have those values specified
    # here.  Those values may be simply over-ridden by user-selected
    # values.  Parameters that have algorithmic default values have
    # their value set to -1 here.  User-selected values can over-ride
    # them.  If the user doesn't select a value, a final value is
    # determined later on.  Note: Keys are the parameter strins
    # used in radvd.conf file, which do not always match exactly the
    # strings used in RFC-4861.

    my %prefix_param_hash = ( 'AdvValidLifetime' => '2592000',
                              'AdvOnLink' => 'on',
                              'AdvPreferredLifetime' => '-1',
                              'AdvAutonomous' => 'on' );

    my @prefix_params = $config->listNodes("$param_root prefix $prefix");
    log_msg("prefix_params for prefix $prefix: @prefix_params\n");

    # Read in parameters set by user
    foreach my $prefix_param (@prefix_params) {
        log_msg("prefix_param = $prefix_param\n");

        my $value = 
            $config->returnValue("$param_root prefix $prefix $prefix_param");

        if ($prefix_param eq "on-link-flag") {
            if ($value eq "true") {
                $prefix_param_hash{'AdvOnLink'} = 'on';
            } else {
                $prefix_param_hash{'AdvOnLink'} = 'off';
            }
        } elsif ($prefix_param eq "autonomous-flag") {
            if ($value eq "true") {
                $prefix_param_hash{'AdvAutonomous'} = 'on';
            } else {
                $prefix_param_hash{'AdvAutonomous'} = 'off';
            }
        } elsif ($prefix_param eq "valid-lifetime") {
            $prefix_param_hash{'AdvValidLifetime'} = $value;
        } elsif ($prefix_param eq "preferred-lifetime") {
            $prefix_param_hash{'AdvPreferredLifetime'} = $value;
        }
        
    }

    # Fill in any missing default values
    if ($prefix_param_hash{'AdvPreferredLifetime'} == -1) {
        if (($prefix_param_hash{'AdvValidLifetime'} > 604800) ||
            ($prefix_param_hash{'AdvValidLifetime'} eq "infinity")) {
            $prefix_param_hash{'AdvPreferredLifetime'} = 604800;
        } else {
            $prefix_param_hash{'AdvPreferredLifetime'} = 
                $prefix_param_hash{'AdvValidLifetime'};
        }
    }

    # Validate user parameters
    if (($prefix_param_hash{'AdvValidLifetime'} ne "infinity") &&
        (($prefix_param_hash{'AdvPreferredLifetime'} eq "infinity") ||
         ($prefix_param_hash{'AdvPreferredLifetime'} > 
          $prefix_param_hash{'AdvValidLifetime'}))) {
        printf("Error: You have set AdvPreferredLifetime to ");
        printf("$prefix_param_hash{'AdvPreferredLifetime'}\n");
        printf("and AdvValidLifetime to ");
        printf("$prefix_param_hash{'AdvValidLifetime'}.\n");
        printf("AdvPreferredLifetime must be less than or equal to AdvValidLifetime.\n");
        exit(1);
    }

    # Write parameters out to config file
    print $FD_WR "    prefix $prefix {\n";
    foreach my $key (keys %prefix_param_hash) {
        print $FD_WR "        $key $prefix_param_hash{$key};\n";
    }

    print $FD_WR "    };\n";
}

# Parse params and write config file for one interface
sub do_interface {
    my $ifname = shift;
    my $intf = new Vyatta::Interface($ifname);
    die "Unknown interface type $ifname\n" unless $intf;
    my $param_root = $intf->path() . " ipv6 router-advert";
    
    # RFC-4861 parameters and their default values.
    # Parameters with fixed defined default values have those values
    # set here.  Those values may be simply over-ridden by user-selected
    # values.  Parameters that have algorithmic default values
    # have their value set to -1 here.  User-selected values can
    # over-ride them.  If the user doesn't select a value, a final
    # value is determined later on.
    my %param_hash = ( 'AdvSendAdvert' => 'off',
                       'AdvIntervalOpt' => 'off',
                       'MaxRtrAdvInterval' => 600,
                       'MinRtrAdvInterval' => -1,
                       'AdvManagedFlag' => 'off',
                       'AdvOtherConfigFlag' => 'off',
                       'AdvLinkMTU' => '0',
                       'AdvReachableTime' => '0',
                       'AdvRetransTimer' => '0',
                       'AdvCurHopLimit' => '64',
                       'AdvDefaultLifetime' => '-1',
                       'AdvDefaultPreference' => 'medium' );



    my @params = $config->listNodes($param_root);

    log_msg("params = @params\n");

    # Read in top-level params...
    foreach my $param (@params) {
        log_msg("Node: $param \n");

        my $value = $config->returnValue("$param_root $param");
        log_msg("Value: $value\n");

        if ($param eq "max-interval") {
            $param_hash{'MaxRtrAdvInterval'} = $value;
        } elsif ($param eq "min-interval") {
            $param_hash{'MinRtrAdvInterval'} = $value;
        } elsif ($param eq "managed-flag") {
            if ($value eq "true") {
                $param_hash{'AdvManagedFlag'} = 'on';
            } else {
                $param_hash{'AdvManagedFlag'} = 'off';
            }
        } elsif ($param eq "other-config-flag") {
            if ($value eq "true") {
                $param_hash{'AdvOtherConfigFlag'} = 'on';
            } else {
                $param_hash{'AdvOtherConfigFlag'} = 'off';
            }
        } elsif ($param eq "send-advert") {
            if ($value eq "true") {
                $param_hash{'AdvSendAdvert'} = 'on';
                $param_hash{'AdvIntervalOpt'} = 'on';
            } else {
                $param_hash{'AdvSendAdvert'} = 'off';
            }
        } elsif ($param eq "link-mtu") {
            $param_hash{'AdvLinkMTU'} = $value;
        } elsif ($param eq "reachable-time") {
            $param_hash{'AdvReachableTime'} = $value;
        } elsif ($param eq "retrans-timer") {
            $param_hash{'AdvRetransTimer'} = $value;
        } elsif ($param eq "cur-hop-limit") {
            $param_hash{'AdvCurHopLimit'} = $value;
        } elsif ($param eq "default-lifetime") {
            $param_hash{'AdvDefaultLifetime'} = $value;
        } elsif ($param eq "default-preference") {
            $param_hash{'AdvDefaultPreference'} = $value;
        } elsif ($param eq "prefix") {
            # Skip for now.  We'll do these later.
        } elsif ($param eq "name-server") {
            # Skip for now. Hack will be inserted at the end
        }
    }

    # Fill in remainig defaults
    if ($param_hash{'MinRtrAdvInterval'} == -1) {
        if ($param_hash{'MaxRtrAdvInterval'} > 9) {
            $param_hash{'MinRtrAdvInterval'} = 
                $param_hash{'MaxRtrAdvInterval'} * 0.33;
            # Round to nearest integer
            $param_hash{'MinRtrAdvInterval'} = 
                sprintf("%.0f", $param_hash{'MinRtrAdvInterval'});
        } else {
            $param_hash{'MinRtrAdvInterval'} = 3;
        }
    }

    if ($param_hash{'AdvDefaultLifetime'} == -1) {
        $param_hash{'AdvDefaultLifetime'} =
            $param_hash{'MaxRtrAdvInterval'} * 3;
    }

    # Validate values
    my $mtu_param = $param_hash{'AdvLinkMTU'};
    if ($mtu_param > 0) {
        my $live_mtu = `ip link show $ifname | grep mtu | awk '{ print \$5 }'`;
        $live_mtu =~ s/\n//;
        if ($live_mtu != $mtu_param) {
            printf("Warning: link-mtu parameter given ($mtu_param) is different from the MTU\n");
            printf("currently being used on $ifname ($live_mtu).\n");
        }
    }

    if ($param_hash{'MaxRtrAdvInterval'} > 1800) {
        printf("Error: MaxRtrAdvInterval value is $param_hash{'MaxRtrAdvInterval'}. It must be 1800 or less.\n");
        exit 1;
    }

    if ($param_hash{'MaxRtrAdvInterval'} < 4) {
        printf("Error: MaxRtrAdvInterval valueis $param_hash{'MaxRtrAdvInterval'}. It must be 4 or more\n");
        exit 1;
    }

    if ($param_hash{'MinRtrAdvInterval'} < 3) {
        printf("Error: MinRtrAdvInterval value is $param_hash{'MinRtrAdvInterval'}. It must be 3 or more\n");
        exit 1;
    }
    
    if ($param_hash{'MinRtrAdvInterval'} >
        ($param_hash{'MaxRtrAdvInterval'} * 0.75)) {
        printf("Error: MinRtrAdvInterval is $param_hash{'MinRtrAdvInterval'} and MaxRtrAdvInterval is $param_hash{'MaxRtrAdvInterval'}.\n");
        printf("MinRtrAdvInterval must be no greater than 3/4 MaxRtrAdvInterval\n");
        exit 1;
    }

    if (($param_hash{'AdvDefaultLifetime'} != 0) &&
        ($param_hash{'AdvDefaultLifetime'} <
         $param_hash{'MaxRtrAdvInterval'})) {
        printf("Error: AdvDefaultLifetime is $param_hash{'AdvDefaultLifetime'}");
        printf(" and MaxRtrAdvInterval is $param_hash{'MaxRtrAdvInterval'}.\n");
        printf("AdvDefaultLifetime must equal to or greater than MaxRtrAdvInterval.\n");
        exit 1;
    }

    if ($param_hash{'AdvDefaultLifetime'} > 9000) {
        printf("Error: AdvDefaultLifetime vlaue is $param_hash{'AdvDefaultLifetime'}.  It must be 9000 or less.\n");
        exit 1;
    }

    # Spit out top-level params to config file
    my $date = localtime;
    my $user = getpwuid($<) || 'Unknown';

    print $FD_WR "interface $ifname {\n";
    print $FD_WR "#   This section was automatically generated by the Vyatta\n";
    print $FD_WR "#   configuration sub-system.  Do not edit it.\n";
    print $FD_WR "#\n";
    print $FD_WR "#   Generated by $user on $date\n";
    print $FD_WR "#\n";
    print $FD_WR "    IgnoreIfMissing on;\n";
    foreach my $key (keys %param_hash) {
        print $FD_WR "    $key $param_hash{$key};\n";
    }

    # Process prefix params, if any
    my @prefix_params = $config->listNodes("$param_root prefix");
    log_msg("prefix_params = @prefix_params\n");

    foreach my $prefix (@prefix_params) {
        log_msg("prefix = $prefix\n");
        do_prefix($param_root, $prefix);
    }

    # Process Name Servers
    my @nameservers = $config->returnValues("$param_root name-server");
    log_msg("nameservers = @nameservers\n");
    if (@nameservers) {
        print $FD_WR "    RDNSS ";
        foreach my $nameserver (@nameservers) {
            print $FD_WR "$nameserver ";
        }
        print $FD_WR "{\n    };\n";
    }

    # Process Search lists
    my @dnssl = $config->returnValues("$param_root dnssl");
    log_msg("searchdomain = @dnssl\n");
    if (@dnssl)  {
        print $FD_WR "    DNSSL ";
        foreach my $searchdom (@dnssl) {
            print $FD_WR "$searchdom ";
        }
        print $FD_WR "{\n    };\n";
    }

    # Finish off config file
    print $FD_WR "};\n";
}

sub generate_conf {
    my $ifname = shift;
    log_msg("ifname = $ifname\n");

    # Generate config file for this interface into temporary string buffer
    my $conf_buffer;

    open $FD_WR, '>', \$conf_buffer
        or die "can't open temporary buffer: $!";

    $config = new Vyatta::Config;

    # Generate config file section for interface into temp file
    do_interface($ifname);

    close $FD_WR;

    if ( -e $conf_file ) {
        # Delete old parameter section for interface from global config file
        delete_conf($ifname);
    }

    # Cat newly generated temp file into global config file
    log_msg("copying in tempfile...\n");

    open my $CFG, '>>', $conf_file
        or die "can't open $conf_file: $!";

    print {$CFG} $conf_buffer;

    close $CFG;
}


#
# Delete the configuration information for an interface from the config file.
#
sub delete_conf {
    my $ifname = $_[0];
    log_msg("delete_conf ifname = $ifname\n");

    my $sedcmd = "sed -i -e \'/^interface $ifname {/,/^}/d\' $conf_file";
    log_msg("sedcmd is $sedcmd\n");

    my $exit_code = system($sedcmd);
    log_msg("exit_code is $exit_code\n");
    if ($exit_code != 0) {
        printf("Unable to edit radvd system config file.  Exit code $exit_code\n");
        exit 1;
    }
}

# Local Variables:
# mode: perl
# indent-tabs-mode: nil
# perl-indent-level: 4
# End:


scripts/vyatta-dhcpv6-client.pl
#!/usr/bin/perl
#
# Module: vyatta-dhcpv6-client.pl
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2005-2009 Vyatta, Inc.
# All Rights Reserved.
#
# Author: Bob Gilligan <gilligan@vyatta.com>
# Date: April 2010
# Description: Start and stop DHCPv6 client daemon for an interface.
#
# **** End License ****
#
#

use strict;
use lib "/opt/vyatta/share/perl5/";
use FileHandle;
use Vyatta::Config;
use Getopt::Long;

my $start_flag;		# Start the daemon
my $stop_flag;		# Stop the daemon and delete all config files
my $release_flag;	# Stop the daemon, but leave config file
my $renew_flag;		# Re-start the daemon.  Functionally same as start_flag
my $temp_flag;
my $params_only_flag;
my $ifname;


sub gen_conf_file {
    my ($conffile, $ifname) = @_;

    my $FD_WR = new FileHandle;

    if (!open($FD_WR, ">$conffile")) {
	printf("Can't write config file: $conffile\n");
	exit 1;
    }
    my $date = `date`;
    my $user = `id -un`;
    my $hostname = `hostname`;
    chomp($date);
    chomp($user);

    print $FD_WR "# This file was auto-generated by the Vyatta\n";
    print $FD_WR "# configuration sub-system.  Do not edit it.\n";
    print $FD_WR "\n";
    print $FD_WR "#   Generated on $date by $user\n";
    print $FD_WR "#\n";
    print $FD_WR "interface \"$ifname\" {\n";
    print $FD_WR "        send host-name \"$hostname\";\n";
    print $FD_WR "        send dhcp6.oro 1, 2, 7, 12, 13, 23, 24, 39;\n";
    print $FD_WR "}\n";
}
    

#
# Main Section
#

GetOptions("start" => \$start_flag,
	   "stop" => \$stop_flag,
	   "release" => \$release_flag,
	   "renew" => \$renew_flag,
	   "temporary" => \$temp_flag,
	   "parameters-only" => \$params_only_flag,
	   "ifname=s" => \$ifname,
    );

if ((defined $temp_flag) && (defined $params_only_flag)) {
    printf("Error: --temporary and --parameters-only flags are mutually exclusive.\n");
    exit 1;
}

if (!defined $ifname) {
    printf("Error: Interface name must be specified with --ifname parameter.\n");
    exit 1;
}

my $pidfile = "/var/lib/dhcp3/dhclient_v6_$ifname.pid";
my $leasefile = "/var/lib/dhcp3/dhclient_v6_$ifname.leases";
my $conffile = "/var/lib/dhcp3/dhclient_v6_$ifname.conf";
my $cmdname = "/sbin/dhclient";

if (defined $release_flag) {
    if (! -e $conffile) {
	printf("DHCPv6 client is not configured on interface $ifname.\n");
	exit 1;
    }

    if (! -e $pidfile) {
	printf("DHCPv6 client is already released on interface $ifname.\n");
	exit 1;
    }
}

if (defined $renew_flag) {
    if (! -e $conffile) {
	printf("DHCPv6 client is not configured on interface $ifname.\n");
	exit 1;
    }
}

if (defined $stop_flag || defined $release_flag) {
    # Stop dhclient -6 on $ifname

    printf("Stopping daemon...\n");
    my $output=`$cmdname -6 -nw -cf $conffile -pf $pidfile -lf $leasefile -r $ifname`;
    printf($output);
    
    # Delete files it leaves behind...
    printf("Deleting related files...\n");
    unlink($pidfile);
    unlink($leasefile);
    if (!defined $release_flag) {
	# If just releasing, leave the config file around as a flag that
	# DHCPv6 remains configured on this interface.
	unlink($conffile);
    }
}

if (defined $start_flag || defined $renew_flag) {
    # Generate the DHCP client config file...
    gen_conf_file($conffile, $ifname);

    # First, kill any previous instance of dhclient running on this interface
    #
    printf("Stopping old daemon...\n");
    my $output = `$cmdname -6 -pf $pidfile -x $ifname`;
    printf($output);

    # start "dhclient -6" on $ifname

    my $args = "";
    if (defined $temp_flag) {
	$args .= " -T";
    }
    if (defined $params_only_flag) {
	$args .= " -S";
    }

    printf("Starting new daemon...\n");
    my $output=`$cmdname -6 -nw -cf $conffile -pf $pidfile -lf $leasefile $args $ifname`;
    printf($output);
}
